### PR TITLE
[SPARK-3181][MLLIB]: Add Robust Regression Algorithm with Huber Estimator

### PR DESCRIPTION
**Pull request by fjiang6**
Huber Robust Regression under spark/ml/regression 
Unit Tests

### CODE REVIEW

**Comment by fjiang6** [comment]
@dbtsai @srowen had RobustRegression in the same LinearRegression codebase as requested. And included the Unit Tests

**Comment by fjiang6** [comment]
@mengxr @dbtsai @srowen had RobustRegression in the same LinearRegression codebase as requested. And included the Unit Tests.

**Comment by dbtsai** [comment]
Still a lot of duplication. We're adding new features into LiR now, and it will be hard to maintain. Is it possible that you just add the objective function, and use Params to switch between different objective function? Thanks.

**Comment by fjiang6** [comment]
@dbtsai ust added the objective function, and use Params to switch between different objective function. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala [hunk 1]
```diff
@@ -325,4 +325,21 @@ private[ml] trait HasStepSize extends Params {
   /** @group getParam */
   final def getStepSize: Double = $(stepSize)
 }
+
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: `sharedParams.scala` can not be edited directly. Please look at `SharedParamsCodeGen.scala`.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: Also, make `HasRobust` as `HasRobustRegression` in `SharedParamsCodeGen.scala`.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 2]
```diff
@@ -123,6 +123,14 @@ class LinearRegression(override val uid: String)
   def setTol(value: Double): this.type = set(tol, value)
   setDefault(tol -> 1E-6)
 
+  /**
+   * Set the robust Option.
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Make the document more explicitly.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 3]
```diff
@@ -645,3 +658,80 @@ private class LeastSquaresCostFun(
...
   }
 }
+
+/**
+ * HuberCostFun implements Breeze's DiffFunction[T] for Huber cost as used in Robust regression.
+ * The Huber M-estimator corresponds to a probability distribution for the errors which is normal
+ * in the centre but like a double exponential distribution in the tails (Hogg 1979: 109).
+ * L = 1/2 ||A weights-y||^2 if |A weights-y| <= k
+ * L = k |A weights-y| - 1/2 K^2 if |A weights-y| > k
+ * where k = 1.345 which produce 95% efficiency when the errors are normal and
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Is it possible to make k tunable and default to 1.345 by having another param?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 4]
```diff
@@ -645,3 +658,80 @@ private class LeastSquaresCostFun(
...
+    standardization: Boolean,
+    featuresStd: Array[Double],
+    featuresMean: Array[Double],
+    effectiveL2regParam: Double) extends DiffFunction[BDV[Double]] {
+
+  override def calculate(weights: BDV[Double]): (Double, BDV[Double]) = {
+    val w = Vectors.fromBreeze(weights)
+
+    val leastSquaresAggregator = data.treeAggregate(new LeastSquaresAggregator(w, labelStd,
+      labelMean, fitIntercept, featuresStd, featuresMean))(
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Do you have reference paper of the objective function of robust regression? My gut feeling is that the condition is per instance, which means if the one particular sample has high error, we make the impact smaller. But you are still using `LeastSquaresAggregator`, and have the condition applied in the whole objective function. I may be wrong. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 3 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: BTW, will be nice that users can set % of efficiency instead of 1.345 this kind of magic number.

File: mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala [hunk 5]
```diff
@@ -505,4 +505,455 @@ class LinearRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+  test("Robust regression with intercept without regularization") {
+    val trainer1 = (new LinearRegression).setRobust(true)
+    // The result should be the same regardless of standardization without regularization
+    val trainer2 = (new LinearRegression).setRobust(true).setStandardization(false)
+    val model1 = trainer1.fit(dataset)
+    val model2 = trainer2.fit(dataset)
+
+    /*
+       Using the following R code to load the data and train the model using glmnet package.
+
```
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala]: Why do you test with `GLMNET`? Is it because all the data in this test suite doesn't have outliner? Can you have a test with outliner proving this algorithm works?

**Comment by fjiang6** [comment]
This class was not added by me. I didn't touch PySpark.

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 6]
```diff
@@ -65,6 +65,10 @@ private[shared] object SharedParamsCodeGen {
         isValid = "ParamValidators.inArray(Array(\"skip\", \"error\"))"),
       ParamDesc[Boolean]("standardization", "whether to standardize the training features" +
         " before fitting the model.", Some("true")),
+      ParamDesc[Boolean]("robustRegression", "whether to use robust Huber Cost Function",
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: I think it would be better to introduce a "costFunction" param which defaults to "LeastSquares" and pattern match in LinearRegression#L195 since that will force mutual exclusivity when more than two cost functions are possible

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 7]
```diff
@@ -645,3 +666,179 @@ private class LeastSquaresCostFun(
     (leastSquaresAggregator.loss + regVal, new BDV(totalGradientArray))
   }
 }
+
+/**
+ * HuberCostFun implements Breeze's DiffFunction[T] for Huber cost as used in Robust regression.
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: nit: surround references to code in `[[DiffFunction[T]]]` to have docs generate links

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 8]
```diff
@@ -645,3 +666,179 @@ private class LeastSquaresCostFun(
...
+              val temp = value / (featuresStd(index) * featuresStd(index))
+              totalGradientArray(index) += effectiveL2regParam * temp
+              value * temp
+            } else {
+              0.0
+            }
+          }
+        }
+      }
+      5 * effectiveL2regParam * sum
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Why this is 5?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 9]
```diff
@@ -645,3 +666,179 @@ private class LeastSquaresCostFun(
...
+   * @return This HuberAggregator object.
+   */
+  def add(label: Double, data: Vector): this.type = {
+    require(dim == data.size, s"Dimensions mismatch when adding new sample." +
+      s" Expecting $dim but got ${data.size}.")
+
+    val diff = dot(data, effectiveWeightsVector) - label / labelStd + offset
+    val k = robustEfficiency
+
+    if (diff < -k) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: This looks right now.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 10]
```diff
@@ -645,3 +666,179 @@ private class LeastSquaresCostFun(
...
+    data: RDD[(Double, Vector)],
+    labelStd: Double,
+    labelMean: Double,
+    fitIntercept: Boolean,
+    standardization: Boolean,
+    featuresStd: Array[Double],
+    featuresMean: Array[Double],
+    effectiveL2regParam: Double,
+    robustEfficiency: Double) extends DiffFunction[BDV[Double]] {
+
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Still have lots of duplicated code. Do you think you can pass a parameter to switch different loss?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 11]
```diff
@@ -97,6 +97,22 @@ class LinearRegression(override val uid: String)
   setDefault(standardization -> true)
 
   /**
+   * Set the robust Option to determine whether to use robust Huber Cost Function
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: nit: This is not really an "Option", can we just make this say " Set whether to use robust Huber Cost Function"

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 12]
```diff
@@ -645,3 +666,179 @@ private class LeastSquaresCostFun(
...
+    featuresMean: Array[Double],
+    effectiveL2regParam: Double,
+    robustEfficiency: Double) extends DiffFunction[BDV[Double]] {
+
+  override def calculate(weights: BDV[Double]): (Double, BDV[Double]) = {
+    val w = Vectors.fromBreeze(weights)
+
+    val huberAggregator = data.treeAggregate(new HuberAggregator(w, labelStd,
+      labelMean, fitIntercept, featuresStd, featuresMean, robustEfficiency))(
+        seqOp = (c, v) => (c, v) match {
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: this can just be 
```scala
seqOp = { case (aggregator, (label, features)) => aggregator.add(label, features) }
```
Ditto for `combOp`

**Comment by feynmanliang** [comment]
Please add "Closes #2096" to the bottom of your PR description so the other one gets closed when this merges.

**Comment by feynmanliang** [comment]
There is a lot of code repetition between this and #2096, perhaps you can make the `mllib` one wrap this?

**Comment by dbtsai** [comment]
Hello, robust tuning parameter k should not be a constant as you implemented. 
In the paper, http://users.stat.umn.edu/~sandy/courses/8053/handouts/robust.pdf
`k = 1.345σ` where `σ` is the square error of current weight. But this will be very expensive to compute the current square error of weight and then compute the huber loss, so I think it's reasonable to approximate the square error from previous weight.

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.

@dbtsai there are a few pull requests that were waiting on your review. Can you revisit them even if they are closed?

**Comment by dbtsai** [comment]
@rxin @mengxr I'm back to US from a leave. Going to revisit PRs under me. 

I had worked with @MechCoder to implement Huber estimator in python scikit https://github.com/scikit-learn/scikit-learn/pull/5291 which had been merged. @fjiang6, @MechCoder, @sethah, are you interested in porting this feature to Spark which should be fairly straightforward? 

Thanks.

**Comment by yanboliang** [comment]
@dbtsai I'm interested in porting Huber estimator to Spark. If you did not start it, I can send a PR in a few days. Thanks!

**Comment by dbtsai** [comment]
@yanboliang Sounds great! Thanks.

**Comment by MechCoder** [comment]
I'll be happy to review it.
